rpcclient/roundrobin.go

```diff
@@ -0,0 +1,434 @@
+// Copyright (c) 2015-2018 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package rpcclient
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/btcsuite/websocket"
+)
+
+var (
+	// ErrNoConnAvailable indicates no Conn is available for pick().
+	ErrNoConnAvailable = errors.New("no Connection is available")
+
+	// ErrNoUsableConnAvailable indicates no usable connections available i.e.
+	// client is marked as Shutdown.
+	ErrNoUsableConnAvailable = errors.New("no Connection is available for use, all are marked as Shutdown")
+)
+
+// ConnectionState indicates the current state of connection.
+type ConnectionState int
+
+// String would return a string representation of the said ConnectionState
+func (s ConnectionState) String() string {
+	switch s {
+	case idle:
+		return "IDLE"
+	case connecting:
+		return "CONNECTING"
+	case ready:
+		return "READY"
+	case shutdown:
+		return "SHUTDOWN"
+	default:
+		return "Invalid-State"
+	}
+}
+
+const (
+	// idle indicates the conn is idle.
+	idle ConnectionState = iota
+
+	// connecting indicates the conn is connecting.
+	connecting
+
+	// disconnected indicates the conn marked to be picked by reconnect handler.
+	disconnected
+
+	// ready indicates the conn is ready for work.
+	ready
+
+	// shutdown indicates the ClientConn has started shutting down.
+	shutdown
+)
+
+// HostAddress represents a server the client connects to.
+type HostAddress struct {
+	// Host is the IP address and port of the RPC server you want to connect
+	// to.
+	Host string
+
+	// Endpoint is the websocket endpoint on the RPC server.  This is
+	// typically "ws".
+	Endpoint string
+
+	// retryCount holds the number of times the client has tried to
+	// reconnect to the RPC server.
+	retryCount int64
+}
+
+// Balancer uses Picker to pick connections.
+// It collects and maintains the connectivity states.
+type Balancer interface {
```

Comment:
This should not be exported.  This interface is not directly usable by callers outside of the `rpcclient` package and it pollutes the public API by being exported.